handler/src/main/java/io/netty/handler/ssl/Java9SslUtils.java

```diff
@@ -0,0 +1,229 @@
+        } catch (UnsupportedOperationException ex) {
+            throw ex;
+        } catch (Exception ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    private static void setApplicationProtocols(SSLEngine engine, List<String> supportedProtocols) {
+        SSLParameters parameters = engine.getSSLParameters();
+
+        String[] protocolArray = supportedProtocols.toArray(EmptyArrays.EMPTY_STRINGS);
+        try {
+            SET_APPLICATION_PROTOCOLS.invoke(parameters, new Object[]{protocolArray});
+        } catch (UnsupportedOperationException ex) {
+            throw ex;
+        } catch (Exception ex) {
+            throw new IllegalStateException(ex);
+        }
+        engine.setSSLParameters(parameters);
+    }
+
+    static void setHandshakeApplicationProtocolSelector(
+            SSLEngine engine, BiFunction<SSLEngine, List<String>, String> selector) {
+        try {
+            SET_HANDSHAKE_APPLICATION_PROTOCOL_SELECTOR.invoke(engine, selector);
+        } catch (UnsupportedOperationException ex) {
+            throw ex;
+        } catch (Exception ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    static BiFunction<SSLEngine, List<String>, String> getHandshakeApplicationProtocolSelector(SSLEngine engine) {
+        try {
+            return (BiFunction<SSLEngine, List<String>, String>)
+                    GET_HANDSHAKE_APPLICATION_PROTOCOL_SELECTOR.invoke(engine);
+        } catch (UnsupportedOperationException ex) {
+            throw ex;
+        } catch (Exception ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    private static void installSelector(Java9SslEngine sslEngine,
+                                        JdkApplicationProtocolNegotiator applicationProtocolNegotiator) {
+        checkNotNull(applicationProtocolNegotiator, "applicationProtocolNegotiator");
+
+        setHandshakeApplicationProtocolSelector(sslEngine.getWrappedEngine(),
+                new AlpnSelector(sslEngine, applicationProtocolNegotiator));
+    }
+
+    private static final class AlpnSelector implements BiFunction<SSLEngine, List<String>, String> {
+        private final Java9SslEngine wrappedEngine;
+        private final JdkApplicationProtocolNegotiator applicationNegotiator;
+        private final Set<String> protocols;
+
+        AlpnSelector(Java9SslEngine wrappedEngine, JdkApplicationProtocolNegotiator applicationNegotiator) {
+            this.wrappedEngine = wrappedEngine;
+            this.applicationNegotiator = applicationNegotiator;
+            protocols = Collections.unmodifiableSet(new LinkedHashSet<String>(applicationNegotiator.protocols()));
+        }
+
+        @Override
+        public String apply(SSLEngine sslEngine, List<String> strings) {
+
+            JdkApplicationProtocolNegotiator.ProtocolSelector selector = applicationNegotiator.protocolSelectorFactory()
```

Comment:
nit: consider adding a static import for `ProtocolSelector` to help with formatting.